src/utils.ts

```diff
@@ -21,48 +20,35 @@ import type {
   EnvelopingTxRequest,
   DeployRequest,
   RelayRequest,
-  RelayInfo,
 } from './common';
 import {
   MISSING_SMART_WALLET_ADDRESS,
   MISSING_CALL_FORWARDER,
 } from './constants/errorMessages';
+import RelayClient from './RelayClient';
+import type { HttpClient } from './api/common';
 
 const INTERNAL_TRANSACTION_ESTIMATED_CORRECTION = 20000; // When estimating the gas an internal call is going to spend, we need to substract some gas inherent to send the parameters to the blockchain
 const ESTIMATED_GAS_CORRECTION_FACTOR = 1;
 const SHA3_NULL_S = utils.keccak256('0x');
 
-const selectNextRelay = async (httpClient: HttpClient): Promise<RelayInfo> => {
+const getRelayClientGenerator = function* (httpClient?: HttpClient) {
   const { preferredRelays } = getEnvelopingConfig();
 
-  for (const preferredRelay of preferredRelays ?? []) {
-    let hubInfo;
-    let managerData;
-
-    try {
-      hubInfo = await httpClient.getChainInfo(preferredRelay);
-
-      if (hubInfo.ready) {
-        const relayHub = RelayHub__factory.connect(
-          hubInfo.relayHubAddress,
-          getProvider()
-        );
-        managerData = await relayHub.getRelayInfo(hubInfo.relayManagerAddress);
-
-        return {
-          hubInfo,
-          managerData,
-        };
-      }
-    } catch (error) {
-      log.warn('Failed to getChainInfo from hub', error);
-      continue;
-    }
-  }
+  let index = 0;
 
-  log.error('No more hubs available to select');
+  while (true) {
+    const url = preferredRelays[index];
+    index++;
+    const relayClient = new RelayClient(httpClient, url);
 
-  throw new Error('No more hubs available to select');
+    if (index < preferredRelays.length) {
+      yield { relayClient, status: `OK` };
+    } else {
+      index = 0;
+      yield { relayClient, status: `End of list of preferred relays` };
+    }
```

Comment:
If we want to return a status, I think we should use constants or enum to store them

Reply:
Furthermore, it's not clear to me what's the purpose of the status. I think the end of the returned values already explains what the status includes, doesn't it?